subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/AbstractConfigurationReportTask.java

```diff
@@ -0,0 +1,170 @@
+import org.gradle.api.Incubating;
+import org.gradle.api.Project;
+import org.gradle.api.internal.file.FileResolver;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.reporting.Reporting;
+import org.gradle.api.reporting.SingleFileReport;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.diagnostics.configurations.ConfigurationReports;
+import org.gradle.api.tasks.diagnostics.internal.configurations.model.ConfigurationReportModelFactory;
+import org.gradle.api.tasks.diagnostics.internal.configurations.ConfigurationReportsImpl;
+import org.gradle.api.tasks.diagnostics.internal.configurations.model.ConfigurationReportModel;
+import org.gradle.api.tasks.diagnostics.internal.configurations.renderer.AbstractWritableConfigurationReportRenderer;
+import org.gradle.api.tasks.diagnostics.internal.configurations.renderer.ConsoleConfigurationReportRenderer;
+import org.gradle.api.tasks.diagnostics.internal.configurations.spec.AbstractConfigurationReportSpec;
+import org.gradle.internal.logging.text.StyledTextOutput;
+import org.gradle.internal.logging.text.StyledTextOutputFactory;
+import org.gradle.util.internal.ClosureBackedAction;
+import org.gradle.work.DisableCachingByDefault;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.FileWriter;
+
+/**
+ * Base class for reporting tasks which display information about attributes and related data associated to a variant/configuration.
+ *
+ * This class implements {@link Reporting} to make configuring additional file output formats simple.
+ *
+ * @since 7.5
+ */
+@Incubating
+@DisableCachingByDefault(because = "Produces only non-cacheable console output")
+public abstract class AbstractConfigurationReportTask extends DefaultTask implements Reporting<ConfigurationReports> {
+    private final ConfigurationReports reports;
+
+    @Inject protected abstract ObjectFactory getObjectFactory();
+    @Inject protected abstract StyledTextOutputFactory getTextOutputFactory();
+    @Inject protected abstract FileResolver getFileResolver();
+
+    protected abstract AbstractConfigurationReportSpec buildReportSpec();
+
+    public AbstractConfigurationReportTask() {
+        ObjectFactory objects = getObjectFactory();
+        reports = objects.newInstance(ConfigurationReportsImpl.class, this);
+    }
+
+    /**
+     * The reports to be generated by this task.
+     */
+    @Override
+    @Nested
+    public final ConfigurationReports getReports() {
+        return reports;
+    }
+
+    /**
+     * Configures the reports to be generated by this task.
+     *
+     * The contained reports can be configured by task name and closures.
+     *
+     * For example, to configure the "outgoingVariants" report to emit a JSON file:
+     *
+     * <pre>
+     * outgoingVariants {
+     *   reports {
+     *     json {
+     *       required = true
+     *       outputLocation.set(file("build/reports/outgoingVariants.js")
```

Comment:
If this is kept, this extension should probably be `.json`.
```suggestion
     *       outputLocation.set(file("build/reports/outgoingVariants.json")
```